runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -515,24 +515,43 @@ J9::ClassEnv::getVFTEntry(TR::Compilation *comp, TR_OpaqueClassBlock* clazz, int
 uint8_t *
 J9::ClassEnv::getROMClassRefName(TR::Compilation *comp, TR_OpaqueClassBlock *clazz, uint32_t cpIndex, int &classRefLen)
    {
+   J9ROMConstantPoolItem *romCP = getROMConstantPool(comp, clazz);
+   J9ROMFieldRef *romFieldRef = (J9ROMFieldRef *)&romCP[cpIndex];
 #if defined(JITSERVER_SUPPORT)
-   auto stream = TR::CompilationInfo::getStream();
-   if (stream && comp->isOutOfProcessCompilation())
+   if (comp->isOutOfProcessCompilation())
       {
-      stream->write(JITServer::MessageType::ClassEnv_getROMClassRefName, clazz, cpIndex);
-      auto classRefNameStr = std::get<0>(stream->read<std::string>());
-      classRefLen = classRefNameStr.length();
-      uint8_t *classRefName = (uint8_t *) comp->trMemory()->allocateHeapMemory(classRefLen);
-      memcpy(classRefName, &classRefNameStr[0], classRefLen);
-      return classRefName;
+      TR::CompilationInfoPerThread *compInfoPT = TR::compInfoPT;
+      char *name;
+      OMR::CriticalSection getRemoteROMClass(compInfoPT->getClientData()->getROMMapMonitor()); 
+      auto &classMap = compInfoPT->getClientData()->getROMClassMap();
+      auto it = classMap.find(reinterpret_cast<J9Class *>(clazz));
+      auto &classInfo = it->second;
+      name = classInfo.getROMString(classRefLen, romFieldRef,
```

Comment:
The implementation of this method (`getROMClassRefName`) is very similar to the implementation of `classNameOfFieldOrStatic(I_32 cpIndex, int32_t & len)`. In there, there are two dereferences before applying `getROMString`
```
  J9ROMFieldRef * ref = (J9ROMFieldRef *) (&romCPBase()[cpIndex]);
   TR_ASSERT(inROMClass(ref), "field ref must be in ROM class"); // for now. if this fails make it a remote call
   char *name = getROMString(len, &romCPBase()[ref->classRefCPIndex],
                           {
                           offsetof(J9ROMClassRef, name),
                           });
```
I think you are missing the second dereference:
```
J9ROMClassRef *romClassRef = (J9ROMClassRef *)&romCP[romFieldRef->classRefCPIndex];
name = classInfo.getROMString(classRefLen, romClassRef,
                            {
                             offsetof(J9ROMClassRef, name)
                             });
```
I would also want some asserts that whatever you get from applying `romCP[some index]` is within the ROMClass boundaries.

Reply:
Yeah, you are right, I'm missing a second lookup into constant pool.